libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -321,6 +323,252 @@ void flatShadingMethod(const TangentSpaceMeshInput* input, TangentSpaceMeshOutpu
+
+        uint3* triangles32 = new uint3[mFaceCount];
+        meshopt_remapIndexBuffer((uint32_t*) triangles32, NULL,  mOutVertices.size(), remap.data());
+
+        float3* outPositions = new float3[vertexCount];
+        float2* outUVs = new float2[vertexCount];
+        quatf* outQuats = new quatf[vertexCount];
+
+        for (size_t i = 0; i < vertexCount; ++i) {
+            outPositions[i] = newVertices[i].position;
+            outUVs[i] = newVertices[i].uv;
+            outQuats[i] = newVertices[i].tangentSpace;
+        }
+
+        output->vertexCount = vertexCount;
+        output->positions = outPositions;
+        output->uvs = outUVs;
+        output->tangentSpace = outQuats;
+        output->triangles32 = triangles32;
+        output->triangleCount = mFaceCount;
+    }
+
+private:
+    static MikktspaceImpl* getThis(SMikkTSpaceContext const* context) noexcept {
+        return (MikktspaceImpl*) context->m_pUserData;
+    }
+
+    inline const uint3 getTriangle(int triangleIndex) const noexcept {
+        const size_t tstride = mIsTriangle16 ? sizeof(ushort3) : sizeof(uint3);
+        return mIsTriangle16 ?
+               uint3(*(ushort3*)(pointerAdd(mTriangles, triangleIndex, tstride))) :
+               *(uint3*)(pointerAdd(mTriangles, triangleIndex, tstride));
+    }
+
+    TangentSpaceMeshOutput* mOutput;
+    int const mFaceCount;
+    float3 const* mPositions;
+    size_t const mPositionStride;
+    float3 const* mNormals;
+    size_t const mNormalStride;
+    float2 const* mUVs;
+    size_t const mUVStride;
+    uint8_t const* mTriangles;
+    bool mIsTriangle16;
+
+    std::vector<IOVertex> mOutVertices;
+};
+
+void mikktspaceMethod(TangentSpaceMeshInput const* input, TangentSpaceMeshOutput* output) {
+    MikktspaceImpl impl(input);
+    impl.run(output);
+}
+
+static float3 randomPerp(const float3& n) {
+    float3 perp = cross(n, float3{1, 0, 0});
+    float sqrlen = dot(perp, perp);
+    if (sqrlen <= std::numeric_limits<float>::epsilon()) {
+        perp = cross(n, float3{0, 1, 0});
+        sqrlen = dot(perp, perp);
+    }
+    return perp / sqrlen;
+}
+
+void lengyelMethod(TangentSpaceMeshInput const* input, TangentSpaceMeshOutput* output) {
+    size_t const vertexCount = input->vertexCount;
+    size_t const triangleCount = input->triangleCount;
+    size_t const positionStride = input->positionStride ? input->positionStride : sizeof(float3);
+    size_t const normalStride = input->normalStride ? input->normalStride : sizeof(float3);
+    size_t const uvStride = input->uvStride ? input->uvStride : sizeof(float2);
+    auto const* triangles16 = input->triangles16;
+    auto const* triangles32 = input->triangles32;
+    auto const* positions = input->positions;
+    auto const* uvs = input->uvs;
+    auto const* normals = input->normals;
+
+    std::vector<float3> tan1(vertexCount);
+    std::vector<float3> tan2(vertexCount);
+    memset(tan1.data(), 0, sizeof(float3) * vertexCount);
```

Comment:
Do we need the `memset` here? Could we instead do something like:

```
    std::vector<float3> tan1(vertexCount, float3{0.0f});
    std::vector<float3> tan2(vertexCount, float3{0.0f});
```

Reply:
done.  

Might be there for optimization (copied from SurfaceOrientation), but will remove since I'm not sure.